### PR TITLE
Prometheus: New instant query results view in Explore

### LINKED ISSUE #32475: Prometheus: Show results of instant queries as text, not in table
**What would you like to be added**:
We are currently showing results of instant queries in the table. This way, we split labels into separate columns. Based on [feedback from community](https://twitter.com/pstbrn/status/1375023722516647937) and discussion with @gouthamve, this can is impractical. Reasons why it is impractical:
- If you have many labels on series, the table either gets super wide (because table rows can't wrap) or you have to hide some labels. It is hard to see everything.
- It is hard to copy-paste values and labels to use them in the your next query. You could use filter-for/filter-out functionality of table, but with that, you can do just 1 label at a time. If you know you want to filter for multiple labels, it would require you to use more resources to filter them one after the other (running multiple queries). Moreover, with having label in column, you are not even able to copy-paste label-value pair.
- We are using virtualized table which means that that if you have a lot of rows, you can't search for all labels. Searching for labels is very important and the fact that not all of them are searchable is big problem. 
- Users would prefer "less-magic" and "more-control"

**Why is this needed**:
Improve experience for Prometheus users.

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What this PR does / why we need it**:
Addresses concerns with UI in explorer with instant prometheus results. 
https://github.com/grafana/grafana/issues/32475

This PR adds a new option to the datasources to present a "Raw" prometheus view, which presents prometheus specific UI in addition to the table view and lets users toggle between them. This new view is the default for prometheus instant query results. 

TLDR:
Users want to be able to copy and paste prometheus results, and the table view is almost unusable for large queries, which is pushing people to the vanilla prometheus UI.

**Which issue(s) this PR fixes**:
https://github.com/grafana/grafana/issues/32475
But note this PR does not add any functionality for searching through results using browser search, the component introduced in this PR is virtualized, and thus will not be searchable via native browser search. We will create another issue to deal with making query results searchable, and only deal with the other usability concerns surfaced in that issue.

Fixes https://github.com/grafana/grafana/issues/32475

**Special notes for your reviewer**:
1. The functionality of this PR aims to emulate the behavior of the native prometheus UI in regard to query results, I would suggest any result in grafana that is different (in order, or content) from the native prometheus UI query results is a bug and should be addressed before release.
2. Where this diverges from prometheus is the ability to compare values across series, or between various computations on the same metric in separate queries. While this new UI isn't intended to provide the ability to compare results from more than 10 queries simultaneously, it typically works well for multiple queries in most of my testing. 
3. This should be considered in alpha or beta, I'd love to get more feedback on this and continue iterating, but at this point (as long as I found all the bugs) I'd like to say this is ready to launch if it's helpful in more use-cases then the table view. For example, the new UI works better with high cardinality databases and large queries (especially for results with many labels that result in tables with many columns).

### CODE REVIEW

**Comment by Elfo404** [comment]
>I had to add a single type assertion to the mapStateToProps function in [TableContainer.tsx:L32](https://github.com/grafana/grafana/pull/55815/files#diff-e69d24177299858aa7ffeca5d9dda0363e1b73ffc20a486e5a1e5a8edb63cd3cR32) to assert that the explore[exploreId] isn't undefined, any ideas on how to avoid this?

you may use `const item: ExploreItemState = explore[exploreId]!;` not ideal still, but we do it quite a few times already in explore, that's something we'd want to refactor anyway at some point

**Comment by gtk-grafana** [comment]
Bunch of design updates and tweaks, here's how things are looking now:

image

image

image

image

image

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+          type="button"
+          onClick={onContentClick}
+        >
+          {isExpandedView ? 'Contract' : 'Expand'}
+        </Button>
+        <div>Result series: {items.length}</div>
+      </header>
+
+      <div role={'table'}>
+        {/* Using one component and changing properties wasn't working well with the virtualized list */}
```
Comment (gtk-grafana) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:91]: is this still true?

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+const getRawListContainerStyles = stylesFactory(() => {
+  return {
+    wrapper: css`
+      height: 100%;
+      overflow: scroll;
+    `,
+    mobileWrapper: css`
+      height: 100%;
+      overflow: scroll;
+      .list-item-attribute {
```
Comment (gtk-grafana) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:30]: I don't like this extra specificity we're adding here with the second class, is there another solution?

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+const RawListContainer = (props: RawListContainerProps) => {
+  const { tableResult } = props;
+  let dataFrame = cloneDeep(tableResult);
+  const styles = getRawListContainerStyles();
+  let listRef = useRef<List | null>(null);
+
+  const valueLabels = dataFrame.fields.filter((field) => field.name.includes('Value'));
+  const items = getRawPrometheusListItemsFromDataFrame(dataFrame);
+  const { width } = useWindowSize();
+  const [isExpandedView, setIsExpandedView] = useState(width <= 480 || valueLabels.length > 2);
```
Comment (gtk-grafana) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:58]: there are variables for the screen width, please use them instead of using seemingly arbitrary number

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      return singleLineHeight;
+    }
+    const item = items[itemIndex];
+
+    // Height of 1.5 lines, plus the number of non value attributes times the height of additional lines
+    return 1.5 * singleLineHeight + (Object.keys(item).length - valueLabels.length) * additionalLineHeight;
+  };
+
+  return (
+    <section data-testid={'raw-list-container'}>
```
Comment (gtk-grafana) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:82]: is this needed?

File: public/app/features/explore/PrometheusListView/RawListItem.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+   * Transform the symbols in the dataFrame to uniform strings
+   */
+  const transformCopyValue = (value: string): string => {
+    if (value === '∞') {
+      return '+Inf';
+    }
+    return value;
+  };
+
+  const stringRep = `${__name__}{${attributeValues.map((value) => {
```
Comment (gtk-grafana) [inline public/app/features/explore/PrometheusListView/RawListItem.tsx:110]: comment please, there's a lot going on here

File: public/app/features/explore/PrometheusListView/RawListItem.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+export interface RawListProps {
+  listItemData: instantQueryRawVirtualizedListData;
+  listKey: number;
+  totalNumberOfValues: number;
+  valueLabels?: Field[];
+  isExpandedView: boolean;
+}
+
+export type RawListValue = { key: string; value: string };
+
```
Comment (gtk-grafana) [inline public/app/features/explore/PrometheusListView/RawListItem.tsx:22]: delete line 22

File: public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.test.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+          type: FieldType.string,
+          values: ['ALERTS', 'ALERTS', 'ALERTS', 'ALERTS_FOR_STATE', 'ALERTS_FOR_STATE', 'ALERTS_FOR_STATE'],
+        },
+        { display, name: 'Value', type: FieldType.number, values: [1, 2, 3, 4, 5, 6] },
+        { display, name: 'attribute', type: FieldType.number, values: [7, 8, 9, 10, 11, 12] },
+      ],
+    });
+    const result = getRawPrometheusListItemsFromDataFrame(dataFrame);
+    result.forEach((row) => {
+      expect(parseInt(row.attribute, 10)).toEqual(parseInt(row.Value, 10) + 6);
```
Comment (gtk-grafana) [inline public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.test.ts:32]: clean this up please

File: public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      const label = field.name;
+
+      if (label !== 'Time') {
+        // Initialize the objects
+        if (typeof field?.display === 'function') {
+          const stringValue = formattedValueToString(field?.display(field.values.get(i)));
+          if (stringValue) {
+            formattedMetric[label] = stringValue;
+          } else if (label.includes('Value #')) {
+            formattedMetric[label] = ' ';
```
Comment (gtk-grafana) [inline public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.ts:43]: can we make this a const and replace all usages?

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 1 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx]: It was kinda true, but we were able to solve this by storing a reference to the virtual list, which provides a callback to render based on style changes.

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 2 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx]: Yeah I moved the prop into the component so it could handle this style

**Comment by gtk-grafana** [comment]
Marking as ready for review. Thanks to @ivanahuckova, @obetomuniz, @diegoadams, @leeoniya, @Elfo404, and more for all your help with this project so far. I think there's still plenty we can do to improve the exploration UX for prometheus, but I think this is in a good place to get this out and see if this is useful for folks.

**Comment by gelicia** [comment]
I'm not familiar with the technicalities of this, but a few UX things

1. Maybe we'd want to keep the default to "table" so this doesn't change what people would see?
2. The "raw" view adds a lot of vertical blank space that snaps back when table view is clicked
3. Does the panel name changing for this make sense? We generally have those toggles be lighter changes, like filtering or something, not something that would alter the name. Would it make sense to draw both the table and the raw view in separate panels?  
4. I wonder if it would make sense to have a setting with a sensible default on what "large" is for prometheus so we can optionally not show a table that doesn't function well if it makes sense.

**Comment by gtk-grafana** [comment]
Hey @gelicia thanks for taking a look at this so quickly!
1. I would argue that the table view is not as helpful for prometheus results in the majority situations (especially for new users to prometheus/grafana), and while I don't want to remove the table view entirely, as there are still a few use cases where the table might work better, I think we want to change the default that people see to be the new view for prometheus instant queries. Also the table view will take longer to load and return visible results to the user. 
2. I'm not sure that I understand, when I toggle between the table and new view the page height stays the same?
3. Putting them in separate panels would be fine, as long as we're only rendering one at a time (we don't want to wait for the table to load). Would this be a blocker for launch, or something we can iterate on after releasing the MVP?
4. Agreed, it would be nice to have some safeguards in the table view to at least display some warnings when we're rendering a certain number of DOM nodes, probably when the table is a certain width/height.

**Comment by gelicia** [comment]
> Hey @gelicia thanks for taking a look at this so quickly!
> 
> 1. I would argue that the table view is not as helpful for prometheus results in the majority situations (especially for new users to prometheus/grafana), and while I don't want to remove the table view entirely, as there are still a few use cases where the table might work better, I think we want to change the default that people see to be the new view for prometheus instant queries. Also the table view will take longer to load and return visible results to the user.
> 2. I'm not sure that I understand, when I toggle between the table and new view the page height stays the same?
> 3. Putting them in separate panels would be fine, as long as we're only rendering one at a time (we don't want to wait for the table to load). Would this be a blocker for launch, or something we can iterate on after releasing the MVP?
> 4. Agreed, it would be nice to have some safeguards in the table view to at least display some warnings when we're rendering a certain number of DOM nodes, probably when the table is a certain width/height.

First off, none of this is a blocker for launch so with 3 and 4, the warnings and separate panels stuff can be evaluated by your team for if it would be helpful or not. 

Good to know that raw will be preferred by most users - I agree then with having it stay default. 

For the page height, see my video

https://user-images.githubusercontent.com/1533128/209692832-a7a25ef0-87a8-4b4f-99cb-b379d10fff26.mov

**Comment by gtk-grafana** [comment]
@gelicia thanks for the video, yeah when there's only a few values it still reserves space to render the entire virtual list. I've played around with this locally a bit, and here's the two solutions I got working locally (github won't let me upload images right now, so I'll send them to you via slack for more context).

I can iterate through the items in short lists and calculate how much space we need to reserve for the virtual list, and since we re-render the virtual list when people switch between expanded/collapsed views, we can dynamically update the height of that component. But then the height of the page suddenly changes, which is kinda jarring and usually places the results below your screen.

Proposals:
1. We dynamically update the height of this view for short lists, which could render content below the viewport for some users
2. We always reserve the max height of the visible elements, this would mean that this table is a bit taller when you load in, but it won't be the whole 600px like in the video you uploaded.

**Review by gelicia — APPROVED** [review]

**Comment by gelicia** [comment]
From an Explore and UI standpoint looks good :D

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    // Height of 1.5 lines, plus the number of non-value attributes times the height of additional lines
+    return 1.5 * singleLineHeight + (Object.keys(item).length - valueLabels.length) * additionalLineHeight;
+  };
+
+  return (
+    <section>
+      <header className={styles.header}>
+        <Button
+          variant="secondary"
+          title="Add query"
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:106]: is this  the title we actually want for the button?

another thing regarding this button is that it lacks some accessibility requirements. i'm not 100% sure the expanded state here is actually expaning/hiding anything, but maybe we should consider it as a toggle button if it is toggling some behaviour: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#associated_aria_roles_states_and_properties

File: public/app/features/explore/TableContainer.tsx [hunk 10]
```diff
@@ -18,27 +20,49 @@ interface TableContainerProps {
...
+
+  return { loading, tableResult: result, range };
 }
 
 const connector = connect(mapStateToProps, {});
 
 type Props = TableContainerProps & ConnectedProps<typeof connector>;
 
-export class TableContainer extends PureComponent<Props> {
+export class TableContainer extends PureComponent<Props, TableContainerState> {
```
Comment (Elfo404) [inline public/app/features/explore/TableContainer.tsx:46]: In a previous discussion it was suggested to create a new component the would import the table rather than extending the table itself. why did that change?

we want to avoid the table to become a "dumpster component" for everything that can be eventually rendered as tabular data (~anything), hence `TableContainer` should really only render a container and its table. i'd suggest to move everything else inside another `*Container` component that holds this logic and eventually reuses some bits of this component (if we need to refactor TableContainer to make that happen happy to see that happen as well) pretty much like LogsContainer reuses some bits from the graph panel to show the volume histogram

**Review by Elfo404 — CHANGES_REQUESTED** [review]

File: public/app/features/explore/TableContainer.tsx (hunk 10 above)
Comment (gtk-grafana) [reply public/app/features/explore/TableContainer.tsx]: Sounds good, I split this off into a new RawPrometheusContainer element, let me know what you think!

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 9 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx]: Good catch with the button title, that was just a sloppy copy/paste error. I'm not sure what else to add for the screen reader markup, as this button only changes the display property (block/inline) of the span elements. I added role=button, but let me know if there's something else we can do to help improve the accessibility UX.

File: public/app/features/explore/PrometheusListView/ItemLabels.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  return (
+    <div className={styles.itemLabelsWrap}>
+      <div className={styles.valueNavigationWrapper}>
+        {valueLabels.map((value, index) => {
+          return (
+            <span className={styles.valueNavigation} key={index}>
+              {value.name}
+            </span>
+          );
+        })}
```
Comment (itsmylife) [inline public/app/features/explore/PrometheusListView/ItemLabels.tsx:36]:
```suggestion
        {valueLabels.map((value, index) => (
          <span className={styles.valueNavigation} key={index}>
            {value.name}
          </span>
        ))}
```

Comment (itsmylife) [reply public/app/features/explore/PrometheusListView/ItemLabels.tsx:36]: Also using index as key is not a good idea. I think we can use value.name as index. WDYT?

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        {
+          <>
+            {/* Show the value headings above all the values, but only if we're in the contracted view */}
+            {valueLabels.length > 1 && !isExpandedView && (
+              <ItemLabels valueLabels={valueLabels} expanded={isExpandedView} />
+            )}
+            <List
+              ref={(list) => {
+                listRef.current = list;
+              }}
```
Comment (itsmylife) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:127]: Shouldn't it be like this? `ref={listRef}`

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+                  });
+                }
+
+                return (
+                  <div role="row" style={{ ...style, overflow: 'hidden' }}>
+                    <RawListItem
+                      isExpandedView={isExpandedView}
+                      valueLabels={filteredValueLabels}
+                      totalNumberOfValues={valueLabels.length}
+                      listKey={index}
```
Comment (itsmylife) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:149]: Again here I think index is not a reliable key

File: public/app/features/explore/PrometheusListView/RawListItem.tsx [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        <ItemValues
+          hideFieldsWithoutValues={hideFieldsWithoutValues}
+          totalNumberOfValues={displayLength}
+          values={values}
+        />
+      </div>
+    </>
+  );
+};
+export default RawListItem;
```
Comment (itsmylife) [inline public/app/features/explore/PrometheusListView/RawListItem.tsx:162]: do you mind adding aria tags as well? For better a11y.

**Review by itsmylife — APPROVED** [review]
A couple of comments but so far LGTM

File: public/app/features/explore/PrometheusListView/ItemLabels.tsx (hunk 11 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/ItemLabels.tsx]: Good catch on both points. Fixed and fixed.

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 13 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx]: Fixed!

File: public/app/features/explore/PrometheusListView/RawListItem.tsx (hunk 14 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListItem.tsx:167]: I added labels for the "cells" of this "table", which is how this UI should be consumed by a screen reader AFAIK, I also removed the expand/contract button from the screen reader API as its functionality is completely useless. Am I missing anything?

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 12 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx]: Yeah it should. Fixed!

File: public/app/features/explore/PrometheusListView/RawListItem.tsx (hunk 14 above)
Comment (itsmylife) [reply public/app/features/explore/PrometheusListView/RawListItem.tsx:167]: I meant not only this component but the whole changeset. Let me go through again and see if I will find anything

File: public/app/features/explore/PrometheusListView/RawListItem.tsx (hunk 5 above)
Comment (bohandley) [reply public/app/features/explore/PrometheusListView/RawListItem.tsx:117]: This is interesting! I only had trouble copying the histogram le buckets that added a zero, i.e. 
The copied object with the le is
`access_evaluation_duration_bucket{instance="localhost:3005",job="go-app",le="0.000160"}`
but the successfully queried metric with labels is 
`access_evaluation_duration_bucket{instance="localhost:3005",job="go-app",le="0.00016"}`

Where did that extra zero come from? 😆 

For now though, I agree that leaving out the "le" is better than giving the user a query that would return no data.

**Review by bohandley — APPROVED** [review]
LGTM! This looks fantastic, I added a comment about copying histogram metrics, but overall great job!

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+      overflow: scroll;
+    `,
+    header: css`
+      display: flex;
+      justify-content: space-between;
+      align-items: center;
+      padding: 10px 0;
+    `,
+  };
+});
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:41]: `stylesFactory` is deprecated, and given those styles are static why not just 
```suggestion
const styles = {
  wrapper: css`
    height: 100%;
    overflow: scroll;
  `,
  mobileWrapper: css`
    height: 100%;
    overflow: scroll;
  `,
  header: css`
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 10px 0;
  `,
};
```

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+const numberOfColumnsBeforeExpandedViewIsDefault = 2;
+
+/**
+ * The container that provides the virtualized list to the child components
+ * @param props
+ * @constructor
+ */
+const RawListContainer = (props: RawListContainerProps) => {
+  const { tableResult } = props;
+  let dataFrame = cloneDeep(tableResult);
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:53]:
```suggestion
  const dataFrame = cloneDeep(tableResult);
```

`dataFrame` is never reassigned

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+  const styles = getRawListContainerStyles();
+  let listRef = useRef<List | null>(null);
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:55]:
```suggestion
  const listRef = useRef<List | null>(null);
```

`listRef` is never reassigned

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 9 above)
Comment (Elfo404) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx]: i see the button has now aria-hidden=true, this makes it practically inaccessible for screen readers users, we should remove it.

also, `role="button"` is implicit for buttons.

my suggestion was mostly using `aria-pressed` to indicate screen readers this is a toggle button:
Screenshot 2022-12-30 at 10 12 47


the issue wit that is that changing content makes screen readers (voice over in this case) skip reading the new state:

https://user-images.githubusercontent.com/1170767/210060183-9a64eee6-e472-466f-b645-d30089e4422e.mov


while preserving the content between renders and only use the icon as an indicator of its state for non-screen reader users works nicely (but is less than ideal):
https://user-images.githubusercontent.com/1170767/210060086-7b902dc4-c4c7-4622-80a5-87fbe126401e.mov


Another option (probably the best), which is also more in line with what is being done in the logs panel for instance, is to use a switch:
https://user-images.githubusercontent.com/1170767/210060728-ccc28be8-54c7-4ec1-92d7-ebb7b8952d50.mov

```tsx
        <Field label="Expand results">
          <Switch value={isExpandedView} onChange={onContentClick} id="someid" />
        </Field>
```

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 18]
```diff
@@ -0,0 +1,162 @@
+import { css } from '@emotion/css';
+import { cloneDeep } from 'lodash';
+import React, { useEffect, useRef, useState } from 'react';
+import { useWindowSize } from 'react-use';
+import { VariableSizeList as List } from 'react-window';
+
+import { DataFrame, Field } from '@grafana/data/src';
+import { Button, stylesFactory } from '@grafana/ui/src';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:8]:
```suggestion
import { DataFrame, Field } from '@grafana/data';
import { Button, stylesFactory } from '@grafana/ui';
```

File: public/app/features/explore/PrometheusListView/RawListItem.tsx [hunk 19]
```diff
@@ -0,0 +1,167 @@
+import { css } from '@emotion/css';
+import React from 'react';
+import { useCopyToClipboard } from 'react-use';
+
+import { Field, GrafanaTheme } from '@grafana/data/src';
+import { IconButton, useStyles } from '@grafana/ui/src';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListItem.tsx:6]:
```suggestion
import { Field, GrafanaTheme } from '@grafana/data';
import { IconButton, useStyles } from '@grafana/ui';
```

File: public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx [hunk 20]
```diff
@@ -0,0 +1,60 @@
+import { css } from '@emotion/css';
+import React from 'react';
+
+import { GrafanaTheme } from '@grafana/data/src';
+import { useStyles } from '@grafana/ui/src';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx:5]:
```suggestion
import { GrafanaTheme } from '@grafana/data';
import { useStyles } from '@grafana/ui';
```

File: public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.test.ts [hunk 21]
```diff
@@ -0,0 +1,36 @@
+import { DataFrame, FieldType, FormattedValue, toDataFrame } from '@grafana/data/src';
```
Comment (Elfo404) [inline public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.test.ts:1]:
```suggestion
import { DataFrame, FieldType, FormattedValue, toDataFrame } from '@grafana/data';
```

File: public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.ts [hunk 22]
```diff
@@ -0,0 +1,60 @@
+import { DataFrame, formattedValueToString } from '@grafana/data/src';
```
Comment (Elfo404) [inline public/app/features/explore/utils/getRawPrometheusListItemsFromDataFrame.ts:1]:
```suggestion
import { DataFrame, formattedValueToString } from '@grafana/data';
```

File: public/app/features/explore/PrometheusListView/ItemLabels.tsx [hunk 23]
```diff
@@ -0,0 +1,38 @@
+import { css } from '@emotion/css';
+import React from 'react';
+
+import { Field } from '@grafana/data/src';
+import { stylesFactory, useStyles } from '@grafana/ui/src';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/ItemLabels.tsx:5]:
```suggestion
import { Field } from '@grafana/data';
import { stylesFactory, useStyles } from '@grafana/ui';
```

File: public/app/features/explore/PrometheusListView/ItemValues.tsx [hunk 24]
```diff
@@ -0,0 +1,80 @@
+import { css } from '@emotion/css';
+import React from 'react';
+
+import { GrafanaTheme } from '@grafana/data/src';
+import { useStyles } from '@grafana/ui/src';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/ItemValues.tsx:5]:
```suggestion
import { GrafanaTheme } from '@grafana/data';
import { useStyles } from '@grafana/ui';
```

File: public/app/features/explore/PrometheusListView/RawListContainer.test.tsx [hunk 25]
```diff
@@ -0,0 +1,70 @@
+import { render, screen, within } from '@testing-library/react';
+import React from 'react';
+
+import { FieldType, FormattedValue, toDataFrame } from '@grafana/data/src';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.test.tsx:4]:
```suggestion
import { FieldType, FormattedValue, toDataFrame } from '@grafana/data';
```

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+          icon={isExpandedView ? 'minus' : 'plus'}
+          type="button"
+          onClick={onContentClick}
+        >
+          {isExpandedView ? 'Contract' : 'Expand'}
+        </Button>
+        <div>Result series: {items.length}</div>
+      </header>
+
+      <div role={'table'}>
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListContainer.tsx:118]: i see you are using divs with roles table/row/cell. what is the advantage over using plain `table`, `tr`s and `td`s? can't we just use them instead?

File: public/app/features/explore/PrometheusListView/ItemLabels.test.tsx [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+  const charactersLength = characters.length;
+  for (let i = 0; i < length; i++) {
+    result += characters.charAt(Math.floor(Math.random() * charactersLength));
+  }
+  return result;
+}
+
+const value1 = getMockString(8);
+const value2 = getMockString(8);
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/ItemLabels.test.tsx:20]: randomizing the test input doesn't really add any value in this case and makes the test unnecessarily complex.
if this was intended to try to catch edge-case values we should probably add specific tests for them instead.

can we remove this and just use contsant values?

File: public/app/features/explore/PrometheusListView/ItemLabels.test.tsx [hunk 28]
```diff
@@ -0,0 +1,60 @@
+import { render, screen } from '@testing-library/react';
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/ItemLabels.test.tsx:1]: This test is exactly the same as `ItemValues.test.tsx`. is it needed?

File: public/app/features/explore/PrometheusListView/ItemValues.test.tsx [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+  const charactersLength = characters.length;
+  for (let i = 0; i < length; i++) {
+    result += characters.charAt(Math.floor(Math.random() * charactersLength));
+  }
+  return result;
+}
+
+const value1 = getMockString(8);
+const value2 = getMockString(8);
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/ItemValues.test.tsx:20]: randomizing the test input doesn't really add any value in this case and makes the test unnecessarily complex.
if this was intended to try to catch edge-case values we should probably add specific tests for them instead.

can we remove this and just use constant values?

File: public/app/plugins/datasource/prometheus/result_transformer.test.ts [hunk 30]
```diff
@@ -131,7 +138,7 @@ describe('Prometheus Result Transformer', () => {
       expect(series.data[0].fields[1].name).toEqual('label1');
       expect(series.data[0].fields[2].name).toEqual('label2');
       expect(series.data[0].fields[3].name).toEqual('Value');
-      expect(series.data[0].meta?.preferredVisualisationType).toEqual('table');
+      expect(series.data[0].meta?.preferredVisualisationType).toEqual('rawPrometheus' as PreferredVisualisationType);
```
Comment (Elfo404) [inline public/app/plugins/datasource/prometheus/result_transformer.test.ts:141]: let's avoid unneeded type assertions (same thing for other changes in this file)

```suggestion
      expect(series.data[0].meta?.preferredVisualisationType).toEqual('rawPrometheus');
```

**Review by Elfo404 — CHANGES_REQUESTED** [review]
I still need to dig deeper, but i've left some comments regarding some top level things in the meantime.

one major thing tho, i see you are using `GrafanaTheme` (and related functions, like `useStyles`) quite extensively, which is deprecated. we should use `GrafanaTheme2` instead.

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx (hunk 26 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx:127]: It's not really a valid table with all the extra markup from the virtual list component, I tried implementing this with our virtual table component, but couldn't get it to provide the functionality we were looking for here, so I stuck with modifying the display of the virtual list component, the data is just tabular enough that I added the extra roles (but mostly to facilitate unit tests).

File: public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+  const styles = useStyles2(getStyles);
+
+  // From the beginning of the string to the start of the `=`
+  const attributeName = value.key;
+
+  // From after the `="` to before the last `"`
+  const attributeValue = value.value;
+
+  return (
+    <span className={isExpandedView ? styles.expanded : styles.contracted} key={index}>
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx:49]:
```suggestion
    <span className={isExpandedView && styles.expanded} key={index}>
```

File: public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+      color: ${greenish};
+    `,
+    metricValue: css`
+      color: ${reddish};
+    `,
+    expanded: css`
+      display: block;
+      text-indent: 1em;
+    `,
+    contracted: css``,
```
Comment (Elfo404) [inline public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx:25]:
```suggestion
```

**Review by Elfo404 — CHANGES_REQUESTED** [review]
Overall looks good. it'd be nice to use more theme tokens instead of hardcoded values if possible. i've left a few more (easy) comments for things i missed when i first reviewed it, if we can fix them then i think we are good to go 🚀

File: public/app/features/explore/PrometheusListView/ItemValues.test.tsx [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+
+export function getMockString(length: number) {
+  let result = '';
+  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+  const charactersLength = characters.length;
+  for (let i = 0; i < length; i++) {
+    result += characters.charAt(Math.floor(Math.random() * charactersLength));
+  }
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/ItemValues.test.tsx:17]: 🤦 fixed

File: public/app/features/explore/PrometheusListView/RawListContainer.tsx [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+        {
+          <>
+            {/* Show the value headings above all the values, but only if we're in the contracted view */}
+            {valueLabels.length > 1 && !isExpandedView && (
+              <ItemLabels valueLabels={valueLabels} expanded={isExpandedView} />
+            )}
+            <List
+              ref={listRef}
+              itemCount={items.length}
+              className={isExpandedView ? styles.mobileWrapper : styles.wrapper}
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListContainer.tsx:137]: No, they were different for a while when I was working on the table solution, this should have been reverted back, good catch!

File: public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx (hunk 31 above)
Comment (gtk-grafana) [reply public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx]: This was throwing a ts error as this className type is `string | false`, I kept the ternary, but deleted that unnecessary contracted css class.

Comment (Elfo404) [reply public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx]: i think you can do `<span className={cx(isExpandedView && styles.expanded)} key={index}>` instead :)

**Review by Elfo404 — APPROVED** [review]
LGTM! 🚀
